src/test_workflow/integ_test/topology.py

```diff
@@ -0,0 +1,109 @@
+        work_dir: Path,
+        component_name: str,
+        additional_cluster_config: dict,
+        bundle_manifest: BundleManifest,
+        security_enabled: bool,
+        component_test_config: str,
+        test_recorder: TestRecorder
+    ) -> None:
+        self.topology_config = topology_config
+        self.dependency_installer = dependency_installer
+        self.work_dir = work_dir
+        self.component_name = component_name
+        self.additional_cluster_config = additional_cluster_config
+        self.bundle_manifest = bundle_manifest
+        self.security_enabled = security_enabled
+        self.component_test_config = component_test_config
+        self.test_recorder = test_recorder
+        self.clusters: list = []
+
+    @classmethod
+    @contextmanager
+    def create(cls, *args: Any) -> Generator[List, None, None]:
+        topology = cls(*args)
+        try:
+            yield topology.create_clusters()
+        finally:
+            topology.destroy()
+
+    def create_clusters(self) -> List:
+        endpoints_list = []
+        original_config = self.additional_cluster_config
+
+        current_port = 9200
+        current_transport_port = 9300
+        current_work_dir_num = 1
+        for cluster_config in self.topology_config:
+
+            data_nodes: list = []
+            manager_nodes: list = []
+            data_nodes_endpoints: list = []
+            manager_nodes_endpoints: list = []
+
+#             Create data nodes
+            for i in range(cluster_config['data_nodes']):
+                cluster_port = current_port
+                original_config['cluster.name'] = "opensearch" + cluster_config['cluster_name']
+                original_config['http.port'] = cluster_port
+                cluster = LocalTestCluster.create_cluster(
+                    self.dependency_installer,
+                    os.path.join(self.work_dir, f'{current_work_dir_num}'),
+                    self.component_name,
+                    original_config,
+                    self.bundle_manifest,
+                    self.security_enabled,
+                    self.component_test_config,
+                    self.test_recorder,
+                    cluster_port)
+                data_nodes.append(cluster)
+                data_nodes_endpoints.append({"endpoint": cluster.endpoint, "port": cluster.port, "transport": current_transport_port})
+
+                current_port = current_port + 1
+                current_work_dir_num = current_work_dir_num + 1
+
+#           Create manager nodes
+#           for i in range (cluster_config['manager_nodes']):
+            self.clusters.append({"cluster_name": cluster_config['cluster_name'], "data_nodes": data_nodes, "manager_nodes": manager_nodes})
+
+            endpoints_list.append(ClusterEndpoint(cluster_config['cluster_name'], data_nodes_endpoints, manager_nodes_endpoints))
+            current_transport_port = current_transport_port + 1
+        return endpoints_list
```

Comment:
This code is very hard to understand (and probably has many bugs). As far as I can read, a cluster is a set of nodes of different type (manager, data). Each node has an endpoint. Each one of these things needs to become a class, like `TestCluster` (already exists) and `Node` with `DataNode` and `ManagerNode` as child classes. This loop would live inside `TestCluster.create(cluster_config)`, iterate over `cluster_config.data_nodes` and call `Node.create(config)` for example, then add the node to `TestCluster.nodes`. The function should return the cluster. If one needs the list of endpoints, call `TestCluster.endpoints` that can iterate over its nodes to collect endpoints.